File: Sources/Client/Pagination.ts
/**
 * Represents information relevant to the pagination of data items.
 */
export class Pagination {

	/**
	 * The current page index.
	 */
	#currentPageIndex!: number;

	/**
	 * The number of items per page.
	 */
	#itemsPerPage!: number;

	/**
	 * The total number of items.
	 */
	#totalItemCount!: number;

	/**
	 * Creates a new pagination.
	 * @param options An object providing values to initialize this instance.
	 */
	constructor(options: PaginationOptions = {}) {
		this.currentPageIndex = options.currentPageIndex ?? 0;
		this.itemsPerPage = options.itemsPerPage ?? 25;
		this.totalItemCount = options.totalItemCount ?? 0;
	}

	/**
	 * The current page index.
	 */
	get currentPageIndex(): number {
		return this.#currentPageIndex;
	}
	set currentPageIndex(value: number) {
		this.#currentPageIndex = Math.max(0, value);
	}

	/**
	 * Value indicating whether a next page exists.
	 */
	get hasNextPage(): boolean {
		return this.#currentPageIndex < this.lastPageIndex;
	}

	/**
	 * Value indicating whether a previous page exists.
	 */
	get hasPreviousPage(): boolean {
		return this.#currentPageIndex > 0;
	}

	/**
	 * The number of items per page.
	 */
	get itemsPerPage(): number {
		return this.#itemsPerPage;
	}
	set itemsPerPage(value: number) {
		this.#itemsPerPage = Math.max(1, Math.min(1000, value));
	}

	/**
	 * The last page index.
	 */
	get lastPageIndex(): number {
		return this.#totalItemCount > 0 ? Math.ceil(this.#totalItemCount / this.#itemsPerPage) - 1 : 0;
	}

	/**
	 * The data limit.
	 */
	get limit(): number {
		return this.#itemsPerPage;
	}

	/**
	 * The data offset.
	 */
	get offset(): number {
		return this.#currentPageIndex * this.#itemsPerPage;
	}

	/**
	 * The total number of items.
	 */
	get totalItemCount(): number {
		return this.#totalItemCount;
	}
	set totalItemCount(value: number) {
		this.#totalItemCount = Math.max(0, value);
	}

	/**
	 * Creates a new pagination from the specified query.
	 * @param query The search parameters providing the query.
	 * @param maxItemsPerPage The maximum number of items allowed per page.
	 * @returns The pagination corresponding to the specified query.
	 */
	static fromQuery(query: string|URLSearchParams, maxItemsPerPage = 1000): Pagination {
		if (typeof query == "string") query = new URLSearchParams(query);

		const parseInt = (key: string, defaultValue: number) => {
			var value = Number(query.get(key) ?? defaultValue.toString());
			return Number.isNaN(value) ? defaultValue : value;
		};

		return new this({
			currentPageIndex: parseInt("Page", 1) - 1,
			itemsPerPage: Math.min(maxItemsPerPage, parseInt("PerPage", 25))
		});
	}
}

/**
 * Defines the options of a {@link Pagination} instance.
 */
export type PaginationOptions = Partial<{

	/**
	 * The current page index.
	 */
	currentPageIndex: number;

	/**
	 * The number of items per page.
	 */
	itemsPerPage: number;

	/**
	 * The total number of items.
	 */
	totalItemCount: number;
}>;

File: Sources/Server/Pagination.cs
namespace Belin.UI;

using System.Globalization;
using System.Text.Json.Serialization;
using System.Web;

/// <summary>
/// Represents information relevant to the pagination of data items.
/// </summary>
public sealed class Pagination {

	/// <summary>
	/// The current page index.
	/// </summary>
	public int CurrentPageIndex { get; set => field = Math.Max(0, value); }

	/// <summary>
	/// Value indicating whether a next page exists.
	/// </summary>
	[JsonIgnore]
	public bool HasNextPage => CurrentPageIndex < LastPageIndex;

	/// <summary>
	/// Value indicating whether a previous page exists.
	/// </summary>
	[JsonIgnore]
	public bool HasPreviousPage => CurrentPageIndex > 0;

	/// <summary>
	/// The number of items per page.
	/// </summary>
	public int ItemsPerPage { get; set => field = Math.Clamp(value, 1, 1000); } = 25;

	/// <summary>
	/// The last page index.
	/// </summary>
	[JsonIgnore]
	public int LastPageIndex => TotalItemCount > 0 ? (int) Math.Ceiling(TotalItemCount / (double) ItemsPerPage) - 1 : 0;

	/// <summary>
	/// The data limit.
	/// </summary>
	[JsonIgnore]
	public int Limit => ItemsPerPage;

	/// <summary>
	/// The data offset.
	/// </summary>
	[JsonIgnore]
	public int Offset => CurrentPageIndex * ItemsPerPage;

	/// <summary>
	/// The total number of items.
	/// </summary>
	public int TotalItemCount { get; set => field = Math.Max(0, value); }

	/// <summary>
	/// Creates a new pagination from the specified query.
	/// </summary>
	/// <param name="query">The dictionary providing the query.</param>
	/// <param name="maxItemsPerPage">The maximum number of items allowed per page.</param>
	/// <returns>The pagination corresponding to the specified query.</returns>
	public static Pagination FromQuery(IDictionary<string, string?> query, int maxItemsPerPage = 1000) {
		int parseInt(string key, int defaultValue) {
			var numericString = query.TryGetValue(key, out var value) ? value : defaultValue.ToString(CultureInfo.InvariantCulture);
			return int.TryParse(numericString, NumberStyles.None, CultureInfo.InvariantCulture, out var result) ? result : defaultValue;
		}

		return new Pagination {
			CurrentPageIndex = parseInt("Page", 1) - 1,
			ItemsPerPage = Math.Min(maxItemsPerPage, parseInt("PerPage", 25))
		};
	}

	/// <summary>
	/// Creates a new pagination from the specified query.
	/// </summary>
	/// <param name="query">The string providing the query.</param>
	/// <param name="maxItemsPerPage">The maximum number of items allowed per page.</param>
	/// <returns>The pagination corresponding to the specified query.</returns>
	public static Pagination FromQuery(string query, int maxItemsPerPage = 1000) {
		var queryString = HttpUtility.ParseQueryString(query);
		return FromQuery(queryString.AllKeys.ToDictionary(key => key ?? "", key => queryString[key]), maxItemsPerPage);
	}
}
